overlord/devicestate/devicemgr.go

```diff
@@ -670,6 +670,10 @@ func (m *DeviceManager) ensureCloudInitRestricted() error {
 		statusMsg := ""
 
 		switch cloudInitStatus {
+		case sysconfig.CloudInitNotFound:
+			// no cloud init at all
+			m.cloudInitAlreadyRestricted = true
```

Comment:
in this case we should disable cloud-init via `DisableCloudInit()` which just drops a config file in /etc which disables cloud-init and doesn't (or shouldn't maybe I have forgotten all the code) rely on the cloud-init binary

Reply:
though we could also just use RestrictCloudInit normally with the special state of CloudInitNotFound. We would likely have to refactor CloudInitState as I recommended however

Reply:
at the moment snaps seems to be calling/requesting cloud-init binary directly, so when it's not there snapd goes into infinite complain loop

Reply:
> at the moment snaps seems to be calling/requesting cloud-init binary directly, so when it's not there snapd goes into infinite complain loop

AFAICT this runs from ensure code, so every ~5 minutes, but may be earlier of EnsureBefore() kicks in for whatever reason, which is what I suspect happened on the device.